File: typescript/math-utils.ts
/**
 * Math transformation utilities (TypeScript).
 *
 * Pure functions — no side effects, no mutation.
 */

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Clamp a value between a lower and upper bound (inclusive). */
export function clamp(value: number, low: number, high: number): number {
  return value < low ? low : value > high ? high : value;
}

/** Linear interpolation between a and b by t [0..1]. */
export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * clamp(t, 0, 1);
}

/** Round a number to a specified number of decimal places. */
export function roundTo(value: number, decimals: number = 0): number {
  const factor = 10 ** decimals;
  return Math.round(value * factor) / factor;
}

/** Calculate the percentage that `part` represents of `total`. */
export function percentage(part: number, total: number): number {
  if (total === 0) return 0;
  return roundTo((part / total) * 100, 2);
}

/** Return the arithmetic mean of an array of numbers. */
export function average(values: number[]): number {
  if (values.length === 0) return 0;
  return values.reduce((sum, v) => sum + v, 0) / values.length;
}

/** Return the median of an array of numbers. */
export function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0
    ? sorted[mid]
    : (sorted[mid - 1] + sorted[mid]) / 2;
}

File: lua/math_utils.lua
--[[
math_utils.lua — Math transformation utilities (Lua)

Pure functions — no side effects, no global state.
--]]

-- ---------------------------------------------------------------------------
-- Clamp
-- ---------------------------------------------------------------------------

--- Clamp a value between a lower and upper bound (inclusive).
function clamp(value, low, high)
  return math.max(low, math.min(high, value))
end

-- ---------------------------------------------------------------------------
-- Lerp
-- ----------------------------------------------------------------------------

--- Linear interpolation between a and b by t [0..1].
function lerp(a, b, t)
  return a + (b - a) * clamp(t, 0, 1)
end

-- ---------------------------------------------------------------------------
-- Round
-- ----------------------------------------------------------------------------

--- Round a number to a specified number of decimal places.
function round(value, decimals)
  decimals = decimals or 0
  local factor = 10 ^ decimals
  return math.floor(value * factor + 0.5) / factor
end

-- ---------------------------------------------------------------------------
-- Average
-- ----------------------------------------------------------------------------

--- Return the arithmetic mean of a table of numbers.
function average(values)
  if #values == 0 then return 0 end
  local sum = 0
  for _, v in ipairs(values) do sum = sum + v end
  return sum / #values
end

-- ---------------------------------------------------------------------------
-- Percentage
-- ----------------------------------------------------------------------------

--- Calculate the percentage that `part` represents of `total`.
function percentage(part, total)
  if total == 0 then return 0 end
  return round((part / total) * 100, 2)
end

return { clamp = clamp, lerp = lerp, round = round, average = average, percentage = percentage }
